examples/api/lib/material/selection_area/selection_area.1.dart

```diff
@@ -0,0 +1,327 @@
+              // Collect spans after selection.
+              afterSelection.add(TextSpan(style: child.style, text: rawText));
+            } else {
+              // Collect spans inside selection.
+              final int newStart = min(endOffset - count, rawText.length);
+              final int globalNewStart = count + newStart;
+              insideSelection.add(TextSpan(style: const TextStyle(color: Colors.red), text: rawText.substring(0, newStart)));
+              // Check if this span contains content after the selection.
+              if (globalNewStart == endOffset && newStart < rawText.length) {
+                afterSelection.add(TextSpan(style: child.style, text: rawText.substring(newStart)));
+              }
+            }
+            count += rawText.length;
+          }
+        } else if (child is WidgetSpan) {
+          if (count < startOffset) {
+            beforeSelection.add(child);
+          } else if (count >= endOffset) {
+            afterSelection.add(child);
+          } else {
+            if (contentRange.children == null) {
+              count += 1;
+              return true;
+            }
+            // Update bulleted list data.
+            for (final SelectedContentRange<Object> range in contentRange.children!) {
+              _colorSelectionRed(
+                <SelectedContentRange<Object>>[range],
+                dataMap: bulletSourceMap,
+                coloringChildSpan: true,
+              );
+            }
+            // Re-create bulleted list.
+            insideSelection.add(
+              WidgetSpan(
+                child: Column(
+                  children: <Widget>[
+                    for (final MapEntry<Key, TextSpan> entry in bulletSourceMap.entries)
+                      Padding(
+                        padding: const EdgeInsets.only(left: 20.0),
+                        child: Text.rich(
+                          bulletSourceMap[entry.key]!,
+                          key: entry.key,
+                        ),
+                      )
+                  ],
+                ),
+              ),
+            );
+          }
+          count += 1;
+        }
+        return true;
+      });
+      dataMap[contentRange.selectableId! as Key] = TextSpan(
+        style: (contentRange.content as TextSpan).style,
+        children: <InlineSpan>[
+          ...beforeSelection,
+          ...insideSelection,
+          ...afterSelection,
+        ],
+      );
+    }
+    // Avoid clearing the selection and setting the state
+    // before we have colored all parts of the selection.
+    if (!coloringChildSpan) {
+      _selectionController.clear();
+      setState(() {});
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+      appBar: AppBar(
+        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
+        title: Text(widget.title),
+      ),
+      body: SelectionArea(
+        controller: _selectionController,
```

Comment:
I was thinking about if there's any way to do this declaratively instead of having a controller with imperative methods on it. I think it's not practical right now, though, without refactoring SelectionArea or something crazy. But for the record I think the declarative way to do this would be:

 * No controller.
 * SelectionArea takes a selectedContentRange parameter. Null means SelectionArea can control its own selection internally, non-null means it must use the selection given by the selectedContentRange.
 * To clear the SelectionArea's selection, clear it in the selectedContentRange and call setState.
 * onSelectionChanged would have to provide the nextSelectedContentRange, and you're responsible for calling setState to pass it back into SelectionArea.

I think that's pretty weird for how SelectionArea works now, so it's not practical to make this declarative, and a controller is fine.

Reply:
If we use declarative we may need to be careful about rebuilds that reset the selection. It will be kinda forcing people to update their input `selectedContentRange` with the `nextSelectedContentRange` when `onSelectionChanged` is called, which may be a error prompt API, and it may not be easily detected (it depends on when the rebuild happens). You may as well force them to have a stateful object like selection controller.

The thing I am hesitate with selection controller is that it may not be easily composable. For example if in the future we want to build a widget that customizes the selection of part of subtree, you may have to do some kind of innerSelectionController and outerSelectionController similar to NestedScrollView, which is quite messy.

I don't have an answer for this as well, just want to point out some pros and cons if we choose one vs the other

Reply:
What if we followed the same approach as `SelectionListener` and made a widget around `SelectionController` (a `SelectionContainer` with a custom delegate that listens to `SelectionController` changes). With this method a user would be able to customize the selection of part of a subtree.

It may also be sufficient to allow the user to use the Shortcuts/Actions system to accomplish clearing the selection.

```dart	
final BuildContext? primaryContext = primaryFocus?.context;
if (primaryContext != null) {
  Actions.invoke(primaryContext, const ClearSelectionIntent());
}
```

Reply:
If we use Actions like that, would it be doable for the user to choose which SelectionArea to clear? Refreshing my own memory: Actions.invoke looks up the tree from primaryContext to find the nearest Actions that handles the given Intent. So to control which SelectionArea, they would need a `context` that's a child of it, with no other nested SelectionArea in between. Is that easier than managing multiple selection controllers? @chunhtai do you have an example of when this gets messy with NestedScrollView to compare?